### ISSUE TITLE
Proto.Actor and .NET Core 3.0

### ISSUE THREAD

**Issue by cpx86** [AsynkronIT/protoactor-dotnet#467]
I recently started looking into what's upcoming in .NET Core 3.0, in particular the new ability to host gRPC services within ASP.NET Core. I did a small request/reply PoC with a normal Proto.Remote client application and an ASP.NET-hosted server application (see https://github.com/cpx86/protoactor-dotnet/tree/aspnetcore3.0-grpc) and while I did get it to work, there are a few things which don't feel natural.

A few thoughts that occured to me that might be interesting to discuss:
1. The biggest pain of all is the way Proto.Remote is designed. `Remote.Start` is not usable at all since it will spin up a separate gRPC server. One thing this highlights is the fact that the concerns of hosting Proto.Remote and of wiring it up are completely mixed into the static `Remote` class. Perhaps we should pick up #411 again and see how we can simplify the design as well as make it natural/easy to use in the .NET Core hosting infrastructure?
2. The static `Proto.Log` class goes very much against the typical way of handling logging, mainly due to the fact that it's static and not injected. As I recall we were never quite happy with this design originally, but went with it because going the injection route might force too much of a change on the rest of the API surface.
3. The `Microsoft.Extensions.*` packages do seem to be part of the "core infrastructure" of .NET Core nowadays. We made the decision some time ago that we did not want to depend on those in Proto.Actor in order to keep it as lightweight as possible. Is this a decision we might want to re-evaluate given how the landscape looks today?

**Comment by alexeyzimarev** [comment]
Injection is the most painful thing in Orleans and I'd suggest not going that path. There was an issue to use LibLog and I personally would rather prefer it, because it adds zero dependencies.

Concerning gRPC, I experimented a bit with it and it was just dying under load. I had the same idea in mind to use what the .NET team has built but the current quality of the solution wasn't satisfactory. At the same time, they use the normal Protobuf and gRPC packages, most of the stuff added is related to the project system (generating C# files) and WebAPI support, so it might be not that much work required to align with .NET Core 3.

**Comment by cpx86** [comment]
Yeah in regards to gRPC my thought was not to build Proto.Remote on top of the ASP.NET Core packages, but rather to refactor it such that it's easier to host Proto in ASP.NET should one wish to do so. To get it working now I basically had to copy-paste the bootstrapping code from `Remote.cs` into `Startup.cs` which is far from ideal.

**Comment by adamhathcock** [comment]
IoC Injection is the pattern being promoted.  I really suggest Proto.Actor cater for it even if it isn't the most performant.  If there's alternate documentation that shows how to avoid it for more perf then so be it.

I wish my focus was still on .NET and this project as I enjoy it but I'm being paid for Typescript and Node 😢

**Comment by alexeyzimarev** [comment]
Well, Microsoft is known for promoting weird patterns at scale for the past two decades. Doesn't mean we need to follow it. Using a static log provider that is wired to any logging library, to obtain the context-specific logger is a widely used pattern as well. DI serves a particular purpose and it has to be used, like any other pattern, where it belongs. I am all pro to have an adapter to build, for example, props by the container. But it must not be the only option like it happened with logging.

**Comment by adamhathcock** [comment]
You're right about LibLog and this library.  I was scared someone was saying something like not supporting actors being created by an IoC container.

Yes, LibLog is a good answer for internal library logging.

**Comment by alexeyzimarev** [comment]
No no, don't get me wrong on that. I just want to have a choice. Like, ASP.NET MVC Core doesn't give me one, like, out of the box. Everything should go to the container. Orleans went that way and I remember a need to create different interfaces where it should be once just to satisfy the container. In MassTransit, for example, we support a lot of containers via separate packages and I still can use hard-written wiring (which is way easier, by the way).

**Comment by jrouaix** [comment]
Well I totally agree with hand written dependency injection being the easiest in many cases.
Do not forget the unit test use case need for IoC #431.

**Comment by cpx86** [comment]
Completely agree on not depending on any specific container or logging infrastructure so users can choose to either hardwire or use a container. Also agree that there is a need and would make sense to refactor the static singletons into a proper IoC pattern.

Going back to `Proto.Remote`, any ideas for how we could improve the current API such that it 1) works well with the new ASP.NET gRPC infrastructure as well as 2) allowing for self-hosting using the built-in gRPC server, preferably without the use of a static singleton `Remote` class?

**Comment by stevehayles** [comment]
Firstly kudos to all concerned, I think this is a fantastic library and although development seems a little stalled it's already very usable.  The comments around the gRPC implementation in .NET Core 3 are very interesting. Having tried to use the new C# version for both the client and server It  definitely feels that a cleaner architecture without static / singletons would help hugely.

I was struggling with the original gRPC implementation as I was having to compile it for an ARM64 environment and couldn't get the native code running reliably. Moving to the .NET Core 3 version of gRPC was harder than I had hoped especially as I was only building a console application and the bootstrapping of gRPC is quiet heavily tied to a full ASP.NET solution 

Not suggesting in any way to move away from gRPC but my final direction lead to me a solution based on RSocket (http://rsocket.io/).  RSocket itself needed lots of work and isn't mature yet on .NET but it was surprisingly easy to integrate.  There is an RSocket RPC spec which effectively re-uses all the Protobuf message contracts so the impact on this library was minimal.

I agree with @cpx86 that it would be very worthwhile looking to improve the current API and without igniting the whole dependency injection argument I think that trying to extract an interface that all these transports share in common might lead to a good solution.  I barely had to touch Remote.cs and the other changes were in EndpointReader.cs and EndpointWriter.cs so it feels like without changing the architecture massively this would be achievable. 

I think the use of Static / Singleton objects is quite well baked into Proto.Actor and will require a more substantial overhaul from someone that knows the codebase but adding flexibility around remoting could be very useful in the long term. 

In summary I am suggesting that if there is a plan to look at more easily integrating the .NET Core 3 version of gRPC then it won't take a lot more to make this pluggable for other options.

**Comment by alexeyzimarev** [comment]
I personally don’t see a big benefit of spreading out dependencies just for the sake of having dependencies. Inevitably, we’ll need the whole dependency management to be baked in and that will force people to use IOC containers. That’s definitely something I would try to avoid.

Potentially, we might go the same way as the OpenTracing library uses. They have the Itracer interface but one can fall back to the GlobalTracer.Instance and use it happily ever after. I’d love to give the possibility for some to have explicit dependencies without forcing everyone else to ditch the static instance.

**Comment by sudsy** [comment]
I was looking at this the other day and was inclined to leave the logging and tracing infrastructure alone at least for the time being.

My plan was to start refactoring Remote.Start from:

```
public static void Start(string hostname, int port, RemoteConfig config)
        {
            RemoteConfig = config;

            ProcessRegistry.Instance.RegisterHostResolver(pid => new RemoteProcess(pid));

            EndpointManager.Start();
            _endpointReader = new EndpointReader();
            _server = new Server
            {
                Services = { Remoting.BindService(_endpointReader) },
                Ports = { new ServerPort(hostname, port, config.ServerCredentials) }
            };
            _server.Start();

            var boundPort = _server.Ports.Single().BoundPort;
            var boundAddr = $"{hostname}:{boundPort}";
            var addr = $"{config.AdvertisedHostname??hostname}:{config.AdvertisedPort?? boundPort}";
            ProcessRegistry.Instance.Address = addr;

            SpawnActivator();

            Logger.LogDebug($"Starting Proto.Actor server on {boundAddr} ({addr})");
        }
```

to something like:

```
public static void Start(string hostname, int port, RemoteConfig config, IRootContext rootContext)
        {
            RemoteConfig = config;

            _endpointReader = new EndpointReader(rootContext);
            _server = new Server
            {
                Services = { Remoting.BindService(_endpointReader) },
                Ports = { new ServerPort(hostname, port, config.ServerCredentials) }
            };
            _server.Start();

            var boundPort = _server.Ports.Single().BoundPort;
            var boundAddr = $"{hostname}:{boundPort}";
            var addr = $"{config.AdvertisedHostname??hostname}:{config.AdvertisedPort?? boundPort}";
         

            Logger.LogDebug($"Starting Proto.Actor server on {boundAddr} ({addr})");
        }
```

The idea being that an EndpointReader service is all that is needed to start a remote and that each EndpointReader instance is linked to a particular RootContext. From what I can see that would satisfy the requirements for starting the service under asp.net core 3.0.

I'm not sure that I have time to do the refactor, but if I found the time it would be good to know I was headed in the right direction. I would appreciate any feedback so that any refactoring efforts would not be in vain.

**Comment by alexeyzimarev** [comment]
When I mentioned OpenTracing, I didn't mean tracing in Proto.Actor. I mean how the OpenTracing .NET library has both an interface that can be passed through as a dependency, and a global instance.

**Comment by alexeyzimarev** [comment]
>  that would satisfy the requirements for starting the service under asp.net core 3.0.

Do you mean that you currently cannot start the service that is using .NET Core 3 runtime? I am a bit confused.

**Comment by sudsy** [comment]
There is now a new way to run grpc under .net core 3. 

[grpc-dotnet](https://github.com/grpc/grpc-dotnet) no longer depends upon the c binaries but does depend upon asp.net core 3. 

As noted on that page, it's not intended to replace the existing method that depends upon the c binaries, it's just presented as an alternative. I will leave the pros and cons of each method for another discussion.

As far as I can tell, there should be no problem running ProtoActor under the .net core 3 runtime as it currently stands. I haven't tried it myself.

**Comment by alexeyzimarev** [comment]
> 2\. The static `Proto.Log` class goes very much against the typical way of handling logging, mainly due to the fact that it's static and not injected.

The injected logger is arguably not the best practice at all. Using static loggers is perfectly fine.  There's a bit of an overhead having it configured everywhere. MassTransit recently solved it by implicitly creating a scoped logger context attached to async local, if a logging operation is called. It allows them to put system diagnostics to the same scope so people can attach diagnostic listeners to exposed diagnostic observables.

**Comment by alexeyzimarev** [comment]
> 3\. The `Microsoft.Extensions.*` packages do seem to be part of the "core infrastructure" of .NET Core nowadays. We made the decision some time ago that we did not want to depend on those in Proto.Actor in order to keep it as lightweight as possible. Is this a decision we might want to re-evaluate given how the landscape looks today?

Well, the dependency on the logging abstractions package is already there.

**Comment by alexeyzimarev** [comment]
Some heads up - for a quick win I removed all the proto generation batch and shell scripts, replaced it with new `Protobuf` item group elements in the csproj files and the Tools package reference. So, protos are generated on save / build.

**Comment by rogeralsing** [comment]
Oh that is very very nice!

**Comment by rogeralsing** [comment]
@alexeyzimarev this can be closed, right?

**Comment by alexeyzimarev** [comment]
@rogeralsing yeah, a lot of the issues mentioned are already addressed. I guess we won't do anything about the logging thing for the moment, but we can improve things if we make something like logging context for actors, combining diagnostic contexts and logging. Maybe give `ActorContext` access to the logger. Something similar was done recently for MassTransit and I did quite a lot of work in different projects with `System.Diagnostics` and activities.
